File: 55.py
from typing import List


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        max_reach = 0
        target = len(nums) - 1

        for i, num in enumerate(nums):
            if i > max_reach:
                return False
            max_reach = max(max_reach, i + num)
            if max_reach >= target:
                return True
        return Fasle

nums1 = [2, 3, 1, 1, 4]
sol = Solution()
print(sol.canJump(nums1))

File: 55.cpp
//
// Created by DuLi on 2023/10/16.
//
#include<iostream>
using std::vector;
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int max_reach = 0;
        int target = nums.size() - 1;

        for (int i = 0; i < nums.size(); ++i) {
            if (i > max_reach){
                return false;
            }
            max_reach = max_reach > (i + nums[i]) ? max_reach : (i + nums[i]);
            if (max_reach >= target){
                return true;
            }
        }
        return false;
    }
};